nav2_costmap_2d/include/nav2_costmap_2d/image.hpp

```diff
@@ -0,0 +1,300 @@
+// Copyright (c) 2021 Andrey Ryzhikov
```

Comment:
You very well could have made this, but this seems suspiciously close to OpenCV. Can you give us a little history on this file? I want to make sure we have the correct attributions here if this was partially or entirely taken from openCV. It will help with tracking down issues later.

Reply:
Is there any of this we can remove or pair down to the bare minimum required to use this layer?  We don't want dead functions or unused code.

Reply:
As well for the image processing library below it

Reply:
About image.hpp. While developing its interface, I knew the OpenCV interface. This definitely influenced me. Just like GDAL, GIL ... The implementation is done without borrowing.
More details:
1. The closest similarity to OpenCV: the `Image` class can both own data and refer to it. To implement this, a pointer to the data and an offset from `i` to `i + 1` row are stored. But this idea is not unique to OpenCV and is widely used in other libraries. 
2. `Image` is templated as opposed to `Mat` from OpenCV.
3. The method names are similar to those of OpenCV. But implemented completely independently. 
Some names are difficult to find a replacement, even on purpose. For example, the choice is between `width` / `columns`, etc. The interface and implementation of many of them differ significantly, for example, the `convert` method is templated and has a custom functor. `forEach` is unique and not present in OpenCV.

Reply:
I can completely remove the use of owning images (p.1) and two methods: `clone`, `part`, since they are only used in tests. But then I will be forced to implement this functionality for tests in the form of functions, not methods. And tests are also code that needs to be maintained. I suppose that this division of functionality is impractical. But it doesn't matter to me.

Reply:
About the file image_processing.hpp. The `MemoryBuffer`, `AsIs`/`BorderConstant` and `Window` classes and related functions are unique. The implementation of morphological operations is fundamentally different from OpenCV (as a result, it is a little slower and, in my opinion, much simpler).

The implementation of the SAUF algorithm (which I heard about from the OpenCV documentation) was made according to an article by its authors. For this reason, it is similar to OpenCV in terms of decision trees and label trees, since exactly this implementation is advised by the article.

Reply:
Got it - removing any / all code that you can from the core implementation that isn't used in core implementation uses is preferable. We don't track test-related code in our metrics and testing-related code I'm significantly more generous about since its not used run-time during execution.

Given the size of this PR, removing as much code there as possible if only being used in tests would be great. It would let me glance over testing code but focus on the run-time code.

Reply:
Dead code removed in [this](https://github.com/ros-planning/navigation2/pull/2567/commits/5e061961895c7e3f4a198af443a55cde4370cbc9) commit.